source/tomopy/misc/corr.py

```diff
@@ -346,6 +348,88 @@ def callback(total, description, unit):
     return arr
 
 
+def median_filter3d(arr, kernel_half_size=1, ncore=None):
+    """
+    Apply 3D median filter to 3D array (C-implementation).
+
+    Parameters
+    ----------
+    arr : ndarray
+        Input 3D array.
+    kernel_half_size : int, optional
+        The half size of the filter's kernel, i.e. 1 results in the full kernel size of 3 x 3 x 3.
+    ncore : int, optional
+        Number of cores that will be assigned to jobs.
+
+    Returns
+    -------
+    ndarray
+        Median filtered 3D array.
+    """
+    input_type = arr.dtype
+    if (input_type != 'float32') and (input_type != 'uint16'):
+        arr = dtype.as_float32(arr) # silently convert to float32 data type
+    out = np.empty_like(arr)
+    dif = 0.0 # set to 0 to avoid selective filtering
+
+    if np.ndim(arr) == 3:
+        dz, dy, dx = arr.shape
+    else:
+        raise ValueError("The input array must be a 3D array")
+    
+    if ncore is None:
+        ncore = mproc.mp.cpu_count()
+        
+    #deal with different data types 
+    if (input_type == 'float32'):
+        extern.c_median_filt3d_float32(arr, out, kernel_half_size, dif, ncore, dx, dy, dz)
+    else:
+        extern.c_median_filt3d_uint16(arr, out, kernel_half_size, dif, ncore, dx, dy, dz)
+    return out
+
+
+def remove_outlier3d(arr, kernel_half_size=1, dif = 0.1, ncore=None):
+    """
+    Also a so-called dezinger. Selectively applies 3D median filter to 
+    3D array (C-implementation) to remove outliers specifically. 
+
+    Parameters
+    ----------
+    arr : ndarray
+        Input 3D array.
+    kernel_half_size : int, optional
+        The half size of the filter's kernel, i.e. 1 results in the full kernel size of 3 x 3 x 3.
+    dif : float
+        Expected difference value between outlier value and
+        the median value of the array.
+    ncore : int, optional
+        Number of cores that will be assigned to jobs.
+
+    Returns
+    -------
+    ndarray
+        Dezingered 3D array.
+    """
+    input_type = arr.dtype
+    if (input_type != 'float32') and (input_type != 'uint16'):
+        arr = dtype.as_float32(arr) # silently convert to float32 data type
+    out = np.empty_like(arr)
+
+    if np.ndim(arr) == 3:
+        dz, dy, dx = arr.shape
+    else:
+        raise ValueError("The input array must be a 3D array")
```

Comment:
0 and 1 are valid dimension lengths for numpy. What happens if any of these dimensions are <= 1?

Reply:
Indeed a good point, I've checked what happens with the function when the lengths are 1, and 0. It doesn't break in both instances and I think we get a 2D filtration in this case (certainly with the length 1). Do you reckon we need to check if we've got a zero length and get a warning?

Reply:
```c
    for(k = 0; k < dimZ; k++)
    {
        for(j = 0; j < dimY; j++)
        {
            for(i = 0; i < dimX; i++)
            {
                index = ((dimX * dimY) * k + j * dimX + i);
                medfilt3D_float(Input, Output, radius, sizefilter_total, mu_threshold, i,
                                j, k, index, (long) (dimX), (long) (dimY), (long) (dimZ));
            }
        }
    }
```
I think when any of `dimZ`, `dimY`, `dimX` are zero, this loop exits before any filtering is done, so I think raising an Exception or UserWarning is needed. I agree that for singleton dimensions the behavior probably devolves to the 2D case.

Reply:
Thanks a lot for the benchmark, looks promising! I removed the 2d implementation inside the code but it can be added very easily. I guess `tomopy.misc.corr.median_filter` could be adapted to it elementary. 

I was wondering, would it make sense making it totally generic then? If `axis` parameter is undefined (e.g. None), then the 3d filter is triggered, when the axis is defined then we have a 2d filter. Or you would still prefer to have a separate instance of median_filter3d? cheers

Reply:
Yes, if we can show performance improvement in both 2D and 3D cases without breaking the existing Python API, then I think adding 3D filtering to `tomopy.median_filter()` instead of adding a new function would be preferred. I think users with existing workflows would appreciate the work-free speedups to their 2D workflows. 😄

Reply:
I've added the 2D version of the filter and it can be called from the `median_filter3d` function with a new parameter `axis` (similar to median_filter but by default  equals to `None`). I've done some basic benchmarking against the TomoPy's multithreading version and it was about two times slower, so I think having `median_filter3d` separately makes more sense. I think I know the reason why it is slower, as I perform slicing in Python and then copy slices to the C code which does 2D filtering. Perform slicing in C (I guess by modifying the filters kernel orientation) should be faster. Could be a task for another day...

Reply:
I appreciate your effort to prototype a new 2D median filter, but since it isn't faster than the existing method, please revert those changes. I don't want to confuse users by providing multiple ways to median filtering with the same method. We can revisit speeding up the 2D filtering in a future PR if you have time.

Reply:
reverted to 3D filtering only.

Reply:
Thanks!